File: Tests/BridgeJSRuntimeTests/JavaScript/ClosureSupportTests.mjs
/**
 * @returns {import('../../../.build/plugins/PackageToJS/outputs/PackageTests/bridge-js.d.ts').Imports["ClosureSupportImports"]}
 */
export function getImports(importsContext) {
    return {
        jsApplyVoid: (fn) => {
            fn();
        },
        jsApplyBool: (fn) => {
            return fn();
        },
        jsApplyInt: (v, fn) => {
            return fn(v);
        },
        jsApplyDouble: (v, fn) => {
            return fn(v);
        },
        jsApplyString: (v, fn) => {
            return fn(v);
        },
        jsApplyJSValue: (v, fn) => {
            return fn(v);
        },
        jsApplyJSObject: (v, fn) => {
            return fn(v);
        },
        jsApplyArrayInt: (v, fn) => {
            return fn(v);
        },
        jsApplyArrayDouble: (v, fn) => {
            return fn(v);
        },
        jsApplyArrayString: (v, fn) => {
            return fn(v);
        },
        jsApplyArrayJSValue: (v, fn) => {
            return fn(v);
        },
        jsApplyArrayJSObject: (v, fn) => {
            return fn(v);
        },
        jsMakeIntToInt: (base) => {
            return (v) => base + v;
        },
        jsMakeDoubleToDouble: (base) => {
            return (v) => base + v;
        },
        jsMakeStringToString: (prefix) => {
            return (name) => `${prefix}${name}`;
        },
        jsCallTwice: (v, fn) => {
            fn(v);
            fn(v);
            return v;
        },
        jsCallBinary: (fn) => fn(1, 2),
        jsCallTriple: (fn) => fn(1, 2, 3),
        jsCallTripleMut: (fn) => {
            fn(1, 2, 3);
            fn(4, 5, 6);
        },
        jsCallTwiceMut: (fn) => {
            fn();
            fn();
        },
        jsHeapCount: () => {
            globalThis.gc?.();
            return globalThis.swift?.memory?._heapValueById?.size ?? -1;
        },
        jsCallAfterRelease: (fn) => {
            try {
                fn();
                fn();
                return "null";
            } catch (e) {
                return e?.message ?? "error";
            }
        },
        jsOptionalInvoke: (fn) => {
            if (fn == null) { return false; }
            return fn();
        },
        jsStoreClosure: (fn) => { globalThis.__storedClosure = fn; },
        jsCallStoredClosure: () => { return globalThis.__storedClosure?.(); },

        runJsClosureSupportTests: () => {
            const exports = importsContext.getExports();
            if (!exports) { throw new Error("No exports!?"); }
            runJsClosureSupportTests(exports);
        },
    };
}

import assert from "node:assert";

/** @param {import('../../../.build/plugins/PackageToJS/outputs/PackageTests/bridge-js.d.ts').Exports} exports */
export function runJsClosureSupportTests(exports) {
    const upperTransform = (text) => text.toUpperCase();
    const processor = new exports.TextProcessor(upperTransform);

    assert.equal(processor.process("hello"), "HELLO");

    const multiParamTransform = (count, text, ratio) => {
        return `${text.toUpperCase()}-${count}-${ratio.toFixed(2)}`;
    };
    assert.equal(processor.processWithCustom("world", multiParamTransform), "WORLD-42-3.14");
    assert.equal(processor.process("test"), "TEST");

    const greeterForClosure = new exports.Greeter("World");
    const greeterCaller = new exports.Greeter("Caller");
    const customGreeting = (greeter) => `Custom greeting for ${greeter.name}: ${greeter.greet()}`;
    const greetResult = greeterCaller.greetWith(greeterForClosure, customGreeting);
    assert.equal(greetResult, "Custom greeting for World: Hello, World!");
    greeterForClosure.release();
    greeterCaller.release();

    assert.equal(exports.formatName("ada", (name) => name.toUpperCase()), "ADA");
    assert.equal(exports.formatName("grace", (name) => `Dr. ${name}`), "Dr. grace");

    const addDr = exports.makeFormatter("Dr.");
    assert.equal(addDr("Ada"), "Dr. Ada");
    assert.equal(addDr("Grace"), "Dr. Grace");

    const addProf = exports.makeFormatter("Prof.");
    assert.equal(addProf("Hopper"), "Prof. Hopper");

    const add10 = exports.makeAdder(10);
    assert.equal(add10(5), 15);
    assert.equal(add10(32), 42);

    const add100 = exports.makeAdder(100);
    assert.equal(add100(23), 123);

    const storedTransform = processor.getTransform();
    assert.equal(storedTransform("hello"), "HELLO");
    assert.equal(storedTransform("world"), "WORLD");

    const greeterForFormatter = new exports.Greeter("Formatter");
    const greeterFormatter = greeterForFormatter.makeFormatter(" [suffix]");
    assert.equal(greeterFormatter("test"), "Hello, Formatter! - test -  [suffix]");
    assert.equal(greeterFormatter("data"), "Hello, Formatter! - data -  [suffix]");
    greeterForFormatter.release();

    const greeterCreator = exports.Greeter.makeCreator("Default");
    const createdG1 = greeterCreator("Alice");
    assert.equal(createdG1.name, "Alice");
    assert.equal(createdG1.greet(), "Hello, Alice!");
    const createdG2 = greeterCreator("");
    assert.equal(createdG2.name, "Default");
    assert.equal(createdG2.greet(), "Hello, Default!");
    createdG1.release();
    createdG2.release();

    const greeterHost = new exports.Greeter("Host");
    const greeterGreeter = greeterHost.makeCustomGreeter();
    const guest1 = new exports.Greeter("Guest1");
    const guest2 = new exports.Greeter("Guest2");
    assert.equal(greeterGreeter(guest1), "Host greets Guest1: Hello, Guest1!");
    assert.equal(greeterGreeter(guest2), "Host greets Guest2: Hello, Guest2!");
    greeterHost.release();
    guest1.release();
    guest2.release();

    const greeterForMethod = new exports.Greeter("Method");
    const greeterParam = new exports.Greeter("Param");
    const methodResult = greeterForMethod.greetWith(greeterParam, (g) => {
        return `Custom: ${g.name} says ${g.greet()}`;
    });
    assert.equal(methodResult, "Custom: Param says Hello, Param!");
    greeterForMethod.release();
    greeterParam.release();

    const optResult1 = processor.processOptionalString((value) => {
        return value !== null ? `Got: ${value}` : `Got: null`;
    });
    assert.equal(optResult1, "Got: test | Got: null");

    const optResult2 = processor.processOptionalInt((value) => {
        return value !== null ? `Number: ${value}` : `Number: null`;
    });
    assert.equal(optResult2, "Number: 42 | Number: null");

    const optResult3 = processor.processOptionalGreeter((greeter) => {
        return greeter !== null ? `Greeter: ${greeter.name}` : `Greeter: null`;
    });
    assert.equal(optResult3, "Greeter: Alice | Greeter: null");

    const optFormatter = processor.makeOptionalStringFormatter();
    assert.equal(optFormatter("world"), "Got: world");
    assert.equal(optFormatter(null), "Got: nil");

    const optCreator = processor.makeOptionalGreeterCreator();
    const opt1 = optCreator();
    assert.equal(opt1, null);
    const opt2 = optCreator();
    assert.notEqual(opt2, null);
    assert.equal(opt2.name, "Greeter2");
    assert.equal(opt2.greet(), "Hello, Greeter2!");
    opt2.release();
    const opt3 = optCreator();
    assert.equal(opt3, null);
    const opt4 = optCreator();
    assert.notEqual(opt4, null);
    assert.equal(opt4.name, "Greeter4");
    opt4.release();

    const dirResult = processor.processDirection((dir) => {
        switch (dir) {
            case exports.Direction.North: return "Going North";
            case exports.Direction.South: return "Going South";
            case exports.Direction.East: return "Going East";
            case exports.Direction.West: return "Going West";
            default: return "Unknown";
        }
    });
    assert.equal(dirResult, "Going North");

    const themeResult = processor.processTheme((theme) => {
        return theme === exports.Theme.Dark ? "Dark mode" : "Light mode";
    });
    assert.equal(themeResult, "Dark mode");

    const statusResult = processor.processHttpStatus((status) => {
        return status;
    });
    assert.equal(statusResult, exports.HttpStatus.Ok);

    const apiResult = processor.processAPIResult((result) => {
        if (result.tag === exports.APIResult.Tag.Success) {
            return `API Success: ${result.param0}`;
        }
        return "API Other";
    });
    assert.equal(apiResult, "API Success: test");

    const dirChecker = processor.makeDirectionChecker();
    assert.equal(dirChecker(exports.Direction.North), true);
    assert.equal(dirChecker(exports.Direction.South), true);
    assert.equal(dirChecker(exports.Direction.East), false);
    assert.equal(dirChecker(exports.Direction.West), false);

    const themeValidator = processor.makeThemeValidator();
    assert.equal(themeValidator(exports.Theme.Dark), true);
    assert.equal(themeValidator(exports.Theme.Light), false);
    assert.equal(themeValidator(exports.Theme.Auto), false);

    const statusExtractor = processor.makeStatusCodeExtractor();
    assert.equal(statusExtractor(exports.HttpStatus.Ok), 200);
    assert.equal(statusExtractor(exports.HttpStatus.NotFound), 404);
    assert.equal(statusExtractor(exports.HttpStatus.ServerError), 500);
    assert.equal(statusExtractor(exports.HttpStatus.Unknown), -1);

    const apiHandler = processor.makeAPIResultHandler();
    assert.equal(apiHandler({ tag: exports.APIResult.Tag.Success, param0: "done" }), "Success: done");
    assert.equal(apiHandler({ tag: exports.APIResult.Tag.Failure, param0: 500 }), "Failure: 500");
    assert.equal(apiHandler({ tag: exports.APIResult.Tag.Info }), "Info");
    assert.equal(apiHandler({ tag: exports.APIResult.Tag.Flag, param0: true }), "Flag: true");
    assert.equal(apiHandler({ tag: exports.APIResult.Tag.Rate, param0: 1.5 }), "Rate: 1.5");
    assert.equal(apiHandler({ tag: exports.APIResult.Tag.Precise, param0: 3.14159 }), "Precise: 3.14159");

    const optDirResult = processor.processOptionalDirection((dir) => {
        return dir !== null ? `Dir: ${dir}` : "Dir: null";
    });
    assert.equal(optDirResult, `Dir: ${exports.Direction.North} | Dir: null`);

    const optThemeResult = processor.processOptionalTheme((theme) => {
        return theme !== null ? `Theme: ${theme}` : "Theme: null";
    });
    assert.equal(optThemeResult, `Theme: ${exports.Theme.Light} | Theme: null`);

    const optApiResult = processor.processOptionalAPIResult((result) => {
        if (result === null) return "Result: null";
        if (result.tag === exports.APIResult.Tag.Success) {
            return `Result: Success(${result.param0})`;
        }
        return "Result: other";
    });
    assert.equal(optApiResult, "Result: Success(ok) | Result: null");

    const optDirFormatter = processor.makeOptionalDirectionFormatter();
    assert.equal(optDirFormatter(exports.Direction.North), "N");
    assert.equal(optDirFormatter(exports.Direction.South), "S");
    assert.equal(optDirFormatter(exports.Direction.East), "E");
    assert.equal(optDirFormatter(exports.Direction.West), "W");
    assert.equal(optDirFormatter(null), "nil");

    processor.release();


    const intToInt = exports.ClosureSupportExports.makeIntToInt(10);
    assert.equal(intToInt(0), 10);
    assert.equal(intToInt(32), 42);

    const doubleToDouble = exports.ClosureSupportExports.makeDoubleToDouble(10.0);
    assert.equal(doubleToDouble(0.0), 10.0);
    assert.equal(doubleToDouble(32.0), 42.0);

    const stringToString = exports.ClosureSupportExports.makeStringToString("Hello, ");
    assert.equal(stringToString("world!"), "Hello, world!");

    const jsIntToInt = exports.ClosureSupportExports.makeJSIntToInt(10);
    assert.equal(jsIntToInt(0), 10);
    assert.equal(jsIntToInt(32), 42);

    const jsDoubleToDouble = exports.ClosureSupportExports.makeJSDoubleToDouble(10.0);
    assert.equal(jsDoubleToDouble(0.0), 10.0);
    assert.equal(jsDoubleToDouble(32.0), 42.0);

    const jsStringToString = exports.ClosureSupportExports.makeJSStringToString("Hello, ");
    assert.equal(jsStringToString("world!"), "Hello, world!");
}

File: Tests/BridgeJSRuntimeTests/ClosureSupportTests.swift
import XCTest
@_spi(Experimental) import JavaScriptKit

@JSClass struct ClosureSupportImports {

    @JSFunction static func jsApplyVoid(_ callback: JSTypedClosure<() -> Void>) throws(JSException)
    @JSFunction static func jsApplyBool(_ callback: JSTypedClosure<() -> Bool>) throws(JSException) -> Bool
    @JSFunction static func jsApplyInt(
        _ value: Int,
        _ transform: JSTypedClosure<(Int) -> Int>
    ) throws(JSException) -> Int
    @JSFunction static func jsApplyDouble(
        _ value: Double,
        _ transform: JSTypedClosure<(Double) -> Double>
    ) throws(JSException) -> Double
    @JSFunction static func jsApplyString(
        _ value: String,
        _ transform: JSTypedClosure<(String) -> String>
    ) throws(JSException) -> String
    // @JSFunction static func jsApplyJSValue(
    //     _ value: JSValue,
    //     _ transform: JSTypedClosure<(JSValue) -> JSValue>
    // ) throws(JSException) -> JSValue
    @JSFunction static func jsApplyJSObject(
        _ value: JSObject,
        _ transform: JSTypedClosure<(JSObject) -> JSObject>
    ) throws(JSException) -> JSObject
    // @JSFunction static func jsApplyArrayInt(
    //     _ value: [Int],
    //     _ transform: JSTypedClosure<([Int]) -> [Int]>
    // ) throws(JSException) -> [Int]
    // @JSFunction static func jsApplyArrayDouble(
    //     _ value: [Double],
    //     _ transform: JSTypedClosure<([Double]) -> [Double]>
    // ) throws(JSException) -> [Double]
    // @JSFunction static func jsApplyArrayString(
    //     _ value: [String],
    //     _ transform: JSTypedClosure<([String]) -> [String]>
    // ) throws(JSException) -> [String]
    // @JSFunction static func jsApplyArrayJSValue(
    //     _ value: [JSValue],
    //     _ transform: JSTypedClosure<([JSValue]) -> [JSValue]>
    // ) throws(JSException) -> [JSValue]
    // @JSFunction static func jsApplyArrayJSObject(
    //     _ value: [JSObject],
    //     _ transform: JSTypedClosure<([JSObject]) -> [JSObject]>
    // ) throws(JSException) -> [JSObject]

    @JSFunction static func jsMakeIntToInt(_ base: Int) throws(JSException) -> (Int) -> Int
    @JSFunction static func jsMakeDoubleToDouble(_ base: Double) throws(JSException) -> (Double) -> Double
    @JSFunction static func jsMakeStringToString(_ `prefix`: String) throws(JSException) -> (String) -> String

    @JSFunction static func jsCallTwice(
        _ value: Int,
        _ callback: JSTypedClosure<(Int) -> Void>
    ) throws(JSException) -> Int
    @JSFunction static func jsCallBinary(_ callback: JSTypedClosure<(Int, Int) -> Int>) throws(JSException) -> Int
    @JSFunction static func jsCallTriple(_ callback: JSTypedClosure<(Int, Int, Int) -> Int>) throws(JSException) -> Int
    @JSFunction static func jsCallAfterRelease(_ callback: JSTypedClosure<() -> Void>) throws(JSException) -> String
    @JSFunction static func jsOptionalInvoke(_ callback: JSTypedClosure<() -> Bool>?) throws(JSException) -> Bool
    @JSFunction static func jsStoreClosure(_ callback: JSTypedClosure<() -> Void>) throws(JSException)
    @JSFunction static func jsCallStoredClosure() throws(JSException)
    @JSFunction static func jsHeapCount() throws(JSException) -> Int

    @JSFunction static func runJsClosureSupportTests() throws(JSException)
}

@JS class ClosureSupportExports {
    @JS static func makeIntToInt(_ base: Int) -> (Int) -> Int {
        return { $0 + base }
    }
    @JS static func makeDoubleToDouble(_ base: Double) -> (Double) -> Double {
        return { $0 + base }
    }
    @JS static func makeStringToString(_ prefix: String) -> (String) -> String {
        return { prefix + $0 }
    }

    @JS static func makeJSIntToInt(_ base: Int) -> JSTypedClosure<(Int) -> Int> {
        return JSTypedClosure { $0 + base }
    }
    @JS static func makeJSDoubleToDouble(_ base: Double) -> JSTypedClosure<(Double) -> Double> {
        return JSTypedClosure { $0 + base }
    }
    @JS static func makeJSStringToString(_ prefix: String) -> JSTypedClosure<(String) -> String> {
        return JSTypedClosure { prefix + $0 }
    }
}

final class ClosureSupportTests: XCTestCase {

    func testRunJsClosureSupportTests() throws {
        try ClosureSupportImports.runJsClosureSupportTests()
    }

    func testClosureParameterVoidToVoid() throws {
        var called = false
        let transform = JSTypedClosure<() -> Void> {
            called = true
        }
        defer { transform.release() }
        try ClosureSupportImports.jsApplyVoid(transform)
        XCTAssertTrue(called)
    }

    func testClosureParameterBoolToBool() throws {
        let transform = JSTypedClosure<() -> Bool> { true }
        defer { transform.release() }
        let result = try ClosureSupportImports.jsApplyBool(transform)
        XCTAssertTrue(result)
    }

    func testClosureParameterIntToInt() throws {
        let transform = JSTypedClosure { $0 * 2 }
        defer { transform.release() }
        let result = try ClosureSupportImports.jsApplyInt(21, transform)
        XCTAssertEqual(result, 42)
    }

    func testClosureParameterDoubleToDouble() throws {
        let transform = JSTypedClosure<(Double) -> Double> { $0 * 2 }
        defer { transform.release() }
        let result = try ClosureSupportImports.jsApplyDouble(21.0, transform)
        XCTAssertEqual(result, 42.0)
    }

    func testClosureParameterStringToString() throws {
        let transform = JSTypedClosure { (value: String) in
            value + ", world!"
        }
        defer { transform.release() }
        let result = try ClosureSupportImports.jsApplyString("Hello", transform)
        XCTAssertEqual(result, "Hello, world!")
    }

    // func testClosureParameterJSValueToJSValue() throws {
    //     let transform = JSTypedClosure<(JSValue) -> JSValue> { $0 }
    //     defer { transform.release() }
    //     let result = try JSTypedClosureImports.jsApplyJSValue(.number(1), transform)
    //     XCTAssertEqual(result, .number(1))
    // }

    func testClosureParameterJSObjectToJSObject() throws {
        let transform = JSTypedClosure<(JSObject) -> JSObject> { $0 }
        defer { transform.release() }
        let obj = JSObject()
        let result = try ClosureSupportImports.jsApplyJSObject(obj, transform)
        XCTAssertEqual(result, obj)
    }

    // func testClosureParameterArrayIntToArrayInt() throws {
    //     let transform = JSTypedClosure<([Int]) -> [Int]> { $0 }
    //     defer { transform.release() }
    //     let result = try ClosureSupportImports.jsApplyArrayInt([1, 2, 3], transform)
    //     XCTAssertEqual(result, [1, 2, 3])
    // }

    // func testClosureParameterArrayDoubleToArrayDouble() throws {
    //     let transform = JSTypedClosure<([Double]) -> [Double]> { $0 }
    //     defer { transform.release() }
    //     let result = try ClosureSupportImports.jsApplyArrayDouble([1.0, 2.0, 3.0], transform)
    //     XCTAssertEqual(result, [1.0, 2.0, 3.0])
    // }

    // func testClosureParameterArrayStringToArrayString() throws {
    //     let transform = JSTypedClosure<([String]) -> [String]> { $0 }
    //     defer { transform.release() }
    //     let result = try ClosureSupportImports.jsApplyArrayString(["a", "b", "c"], transform)
    //     XCTAssertEqual(result, ["a", "b", "c"])
    // }

    // func testClosureParameterArrayJSValueToArrayJSValue() throws {
    //     let transform = JSTypedClosure<([JSValue]) -> [JSValue]> { $0 }
    //     defer { transform.release() }
    //     let result = try ClosureSupportImports.jsApplyArrayJSValue([.number(1), .number(2), .number(3)], transform)
    //     XCTAssertEqual(result, [.number(1), .number(2), .number(3)])
    // }

    // func testClosureParameterArrayJSObjectToArrayJSObject() throws {
    //     let transform = JSTypedClosure<([JSObject]) -> [JSObject]> { $0 }
    //     defer { transform.release() }
    //     let obj1 = JSObject()
    //     let obj2 = JSObject()
    //     let obj3 = JSObject()
    //     let result = try ClosureSupportImports.jsApplyArrayJSObject([obj1, obj2, obj3], transform)
    //     XCTAssertEqual(result, [obj1, obj2, obj3])
    // }

    func testClosureReturnIntToInt() throws {
        let c = try ClosureSupportImports.jsMakeIntToInt(10)
        XCTAssertEqual(c(0), 10)
        XCTAssertEqual(c(32), 42)
    }

    func testClosureReturnDoubleToDouble() throws {
        let c = try ClosureSupportImports.jsMakeDoubleToDouble(10.0)
        XCTAssertEqual(c(0.0), 10.0)
        XCTAssertEqual(c(32.0), 42.0)
    }

    func testClosureReturnStringToString() throws {
        let c = try ClosureSupportImports.jsMakeStringToString("Hello, ")
        XCTAssertEqual(c("world!"), "Hello, world!")
    }

    func testClosureParameterIntToVoid() throws {
        var total = 0
        let callback = JSTypedClosure<(Int) -> Void> { value in
            total += value
        }
        defer { callback.release() }
        let ret = try ClosureSupportImports.jsCallTwice(5, callback)
        XCTAssertEqual(ret, 5)
        XCTAssertEqual(total, 10)
    }

    func testCallingReleasedClosureThrows() {
        let transform = JSTypedClosure<(Int) -> Int> { $0 + 1 }
        transform.release()

        do {
            _ = try ClosureSupportImports.jsApplyInt(41, transform)
            XCTFail("Expected JSException for released closure")
        } catch let error {
            let message = error.thrownValue.object?["message"].string
            XCTAssertNotNil(message)
            XCTAssertTrue(message?.contains(#fileID) ?? false, "message=\(message ?? "nil")")
        }
    }

    func testMultipleArity() throws {
        let sum = JSTypedClosure<(Int, Int) -> Int> { $0 + $1 }
        defer { sum.release() }
        XCTAssertEqual(try ClosureSupportImports.jsCallBinary(sum), 3)
    }

    func testTripleArity() throws {
        let sum = JSTypedClosure<(Int, Int, Int) -> Int> { $0 + $1 + $2 }
        defer { sum.release() }
        XCTAssertEqual(try ClosureSupportImports.jsCallTriple(sum), 6)
    }

    func testOptionalNoneSkipsCall() throws {
        XCTAssertFalse(try ClosureSupportImports.jsOptionalInvoke(nil))
    }

    func testOptionalSomeCalls() throws {
        var called = false
        let closure = JSTypedClosure<() -> Bool> {
            called = true
            return true
        }
        defer { closure.release() }
        XCTAssertTrue(try ClosureSupportImports.jsOptionalInvoke(closure))
        XCTAssertTrue(called)
    }

    func testDropDuringCallThenThrows() throws {
        var closure: JSTypedClosure<() -> Void>! = nil
        closure = JSTypedClosure {
            closure.release()
        }
        let message = try ClosureSupportImports.jsCallAfterRelease(closure)
        XCTAssertTrue(
            message.contains(#fileID),
            "\"\(message)\" does not contain expected file name"
        )
    }

    func testStoredClosureAfterReleaseThrows() throws {
        let closure = JSTypedClosure<() -> Void> {}
        try ClosureSupportImports.jsStoreClosure(closure)
        closure.release()
        XCTAssertThrowsError(try ClosureSupportImports.jsCallStoredClosure())
    }

    func testClosuresDoNotLeakHeapEntries() throws {
        let baseline = try ClosureSupportImports.jsHeapCount()

        for _ in 0..<50 {
            let closure: JSTypedClosure<(Int) -> Void> = JSTypedClosure { _ in }
            _ = try? ClosureSupportImports.jsCallTwice(0, closure)
        }

        // Trigger GC and read heap size from JS side
        let after = try ClosureSupportImports.jsHeapCount()
        XCTAssertEqual(after, baseline, "Heap entry count should return to baseline after GC")
    }

}
